### ISSUE TITLE
Add syntax for abbreviations or acronyms

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#252]
A fairly typical requirement is to define an acronym or abbreviation in text using a tooltip. The target markup to produce in HTML is:

```
<abbr title="HyperText Markup Language">HTML</abbr>
```

and in DocBook:

```
<acronym>HTML</acronym>
```

I propose the following syntax for producing this output, selected because it is both readable and intuitive:

```
(HTML)[HyperText Markup Language]
```

Of course, you aren't going to want to type this over and over again in the text. One solution to the repetition is to use document attributes to hold the inline macro, and refer to it instead by the assigned attribute.

```
= Document Title
:HTML: (HTML)[HyperText Markup Language]

{HTML} 5 is changing the web.
```

This is a good approach because it centralizes the definitions and even allows them to moved to a common file and imported:

```
= Document Title
include::abbreviations.txt[]

{HTML} 5 is changing the web.
```

Since the author selects the attribute names to use, there is no worry about conflicts. (Although, attribute names are not case sensitive). It's also very efficient since it's reusing attribute references (just adding a larger table).

If variable name conflicts are a problem, a qualifier could be used, which would also serve to flag it as an abbreviation.

```
= Document Title
:a-HTML: (HTML)[HyperText Markup Language]
:abbr-CDI: (CDI)[Contexts & Dependency Injection]
:REST-: (REST)[Representational state transfer]

{a-HTML} 5 is changing the web.

{abbr-CDI} brings new flexibility to the Java EE platform.

{REST-} is the founding principle of the web.
```

**Comment by flexik** [comment]
+1

**Comment by shahryareiv** [comment]
+1
The "glossaries" package in Latex can be a good reference for features, in Asciidoc I miss features such as:
-expansion of an abbreviation in the first encounter 
-forced short or long forms \acrlong \acrshort \acrfull

**Comment by FeiWongReed** [comment]
+1

It's very important feature for people producing Docbook

**Comment by robjens** [comment]
Any chance we're going to get a (alphabetical) glossary of terms special section with this? Although I must admit, making one by hand is relatively easy (taking the `abbreviations.txt` idea), but would require manual upkeep (formatting, sorting in alphabetical order) while, in one worldview, this would line up next to the (already auto-generated) ToC being that they're both reference indices (meant for quick lookup) and special sections. Especially abbreviation-heavy organizations would benefit from this as looking up meaning of those without a glossary is a really big pain.

**Comment by mojavelinux** [comment]
To be honest, it will probably exist as an extension before it makes it into core. That's a workflow that is proving to be very effective. It gets a solution in place that people can test out sooner and it gives us the confidence and clarity to know what belongs in core and what works better as an extension. Having said all that, activity drives progress, so the best way to get this moving is to start experimenting!

**Comment by robjens** [comment]
@mojavelinux Fair enough. Would be great to see this brought to the table regardless. Plugin would require the core functionality to be in-place though and I noticed the milestone got shifted back a minor, any difficulties or just a priority thing?

**Comment by mojavelinux** [comment]
The milestones aren't set in stone yet. Expect them to change once I go through all the issues and layout a more concrete roadmap.

**Comment by robjens** [comment]
That's fine, I know who it goes and I'm keeping an eye out for any asciidoctor related projects/changes anyway, checking up every now and then :)

**Comment by shahryareiv** [comment]
I suggest to extend attribute definition pattern by using the unused symbols (probably roles and options) to create variations. For example:
```
:html: HTML
```
The original: {html} => HTML



```
:html%first: HyperText Markup Language ({HTML}) 
```
Internal option: In the first use {html} => HyperText Markup Language (HTML) 



```
:html.expanded: HyperText Markup Language  
```
user-defined roles (inline): [.expanded]{HTML} =>HyperText Markup Language



```
:html.highlight: #{html}# 
```
user-defined roles (block): in the block with [.highlight] role every {html} => *HTML*



```
:html%lang=fa: {html%first}اچ‌تی‌ام‌ال
```
When attribute lang=fa : {HTML}=> اچ‌تی‌ام‌ال



```
:html%lang=fa,first: اچ‌تی‌ام‌ال
```
When attribute lang=fa and it is used for the first time: {HTML}=> اچ‌تی‌ام‌ال HyperText Markup Language (HTML)  



```
:html%sec0%sec1: image:something.png[]
```
all {html} in title level 0 and 1 are going to be substitute with the image


I think a bunch of predefined options that can be helpful are:

- first (first appearance)
- definition (for glossary)
- backend-html (when the backend is html or ... This of course by if-backend things but it might be a short-form
- sec0, ... (when in title of section x)

**Comment by nancym** [comment]
For now I'm doing abbreviations by creating attributes like this:

```
:abbr-imho: pass:[<abbr title="in my humble opinion">IMHO</abbr>]
```

Then using `{abbr-imho}` in my `.adoc` files. I'm posting this here to get your opinion about whether this is a good way to do this. Thank you!

**Comment by shahryareiv** [comment]
I do something like this:

```
:RCT: pass:c,a,r,m,q,p[[.acro]#indexterm2:[RCT]#]
:RCT-abbrev: pass:c,a,r,m,q,p[Randomized Controlled Trial ([.acro]#indexterm2:[RCT]#)]
```
Then I normally use `{RCT}` in the `asciidoc` text. 

For HTML output I have written a script to change the first occurrence of the abbrevation to `{RCT}` in the `asciidoc` file before conversion. 

For Latex output (through DocBook), another script reads the `asciidoc` file to generate a tex file with `\newacronym{...` to be included in Latex file later. Also my customization to DBLatex (for DocBook to Latex conversion) converts `<phrase role="acro"><indexterm><primary>RCT</primary></indexterm>RCT</phrase>` to `\gls{RCT}`

**Comment by endowdly** [comment]
@nancym That method is _exactly_ how I do it now. 

But, why not a macro, @mojavelinux? So we can maintain consistency? 

Something like...

``` 

Here is an abbr:HTML[Hyper-Text Markup Language]! 

It can be resusable like footnotes in our abbr:HTML[]! // Unless it's not defined, then it's just 'HTML'. 
```
